src/compiler/parser.ts

```diff
@@ -2717,6 +2745,31 @@ module ts {
             return expr;
         }
 
+        function parseBindingElementInitializer(inParameter: boolean): Expression {
+            // BindingElement[Yield,GeneratorParameter,Await,AsyncParameter] : 
+            //      [+GeneratorParameter] BindingPattern[?Yield,?Await,GeneratorParameter] Initializer[In]opt
+            //      [+AsyncParameter] BindingPattern[?Yield,?Await,AsyncParameter] Initializer[In]opt
+            //      [~GeneratorParameter,~AsyncParameter] BindingPattern[?Yield,?Await] Initializer[In,?Yield,?Await]opt
+            // SingleNameBinding[Yield,GeneratorParameter,Await,AsyncParameter] : 
+            //      [+GeneratorParameter] BindingIdentifier[Yield] Initializer[In]opt
+            //      [+AsyncParameter] BindingIdentifier[Await] Initializer[In]opt
+            //      [~GeneratorParameter,~AsyncParameter] BindingIdentifier[?Yield,?Await] Initializer[In,?Yield,?Await]opt
+            var explicitFlags = ParserContextFlags.None;
+            var ambientFlags = ParserContextFlags.None;
+            if (!inGeneratorParameterContext() && !inAsyncParameterContext()) {

```

Comment:
GP and AP are mutually exclusive, they shouldn't both be **on** (though it seems we may not report on that until we do a grammar check later). 

We're parsing *Initializer* here. As per the grammar, when either GP or AP are set, we do not pass on the ambient values for Yield or Await. When **neither** GP nor AP are set, we pass on the ambient values.

I may end up reversing my mutual exclusion assertion for better parser behavior. 

As far as explicitly setting `ambientFlags`, the key here is that we don't assume every parameter is ambient. We only want to copy the ambient flags when the grammar says we are passing them.

`setContextParameters` could instead take in `setFlags` and `clearFlags` and we could explicitly clear the flags we don't want, but any time we add a new parser-generated context flag we'd have to update all of the references. As written, a call to `setContextParameters` or `setContextParametersAnd` will roughly mimic the ES6 grammar, e.g.:

```TypeScript
// [+GeneratorParameter] 
if (inGeneratorParameterContext()) {
  // BindingPattern[GeneratorParameter, ?Yield]
  setContextParametersAnd(parseBindingPattern, 
    /*explicitFlags*/ ParserContextFlags.GeneratorParameter, 
    /*ambientFlags*/ ParserContextFlags.Yield);

  // Initializer[In]
  setContextParametersAnd(parseInitializer, 
    /*explicitFlags*/ ParserContextFlags.None, // since we invert [In]
    /*ambientFlags*/ ParserContextFlags.None); 
}

// [~GeneratorParameter]
if (!inGeneratorParameterContext()) {
  // BindingPattern[?Yield]
  setContextParametersAnd(parseBindingPattern, 
    /*explicitFlags*/ ParserContextFlags.None, // DisallowIn?
    /*ambientFlags*/ ParserContextFlags.Yield);

  // Initializer[In, ?Yield]
  setContextParametersAnd(parseInitializer, 
    /*explicitFlags*/ ParserContextFlags.None, // since we invert [In]
    /*ambientFlags*/ ParserContextFlags.Yield); 
}
```